### PR TITLE
Filter on columns

### PR DESCRIPTION
**Pull request by JaapWijnen**
Here's a first attempt at implementing filtering on different columns within a model. Would love to have some feedback :) [issue](https://github.com/vapor/fluent/issues/635) created on the Fluent package

### CODE REVIEW

File: Sources/FluentKit/Query/QueryBuilder.swift [hunk 1]
```diff
@@ -625,113 +671,219 @@ public final class QueryBuilder<Model>
...
+}
+
+public func !=~ <Model, Field>(lhs: KeyPath<Model, Field>, rhs: KeyPath<Model, Field>) -> ModelFilter<Model>
+    where Model: FluentKit.Model, Field: FieldRepresentable, Field.Value: CustomStringConvertible
+{
+    return .init(lhs, .contains(inverse: true, .prefix), rhs)
+}
+
+
+public enum ModelFilter<Model> where Model: FluentKit.Model {
```
Comment (tanner0101) [inline Sources/FluentKit/Query/QueryBuilder.swift]: I think we might want a separate struct for holding field filters. `ModelFilter` is only generic on a single model, but when you're comparing two fields, you may want to compare them across models. 

What about something like this:
```swift
ModelValueFilter<Model>
ModelFieldFilter<Left, Right>
```

File: Sources/FluentKit/Query/QueryBuilder.swift [hunk 2]
```diff
@@ -625,113 +671,219 @@ public final class QueryBuilder<Model>
...
+}
+
+public func <= <Model, Field>(lhs: KeyPath<Model, Field>, rhs: KeyPath<Model, Field>) -> ModelFilter<Model>
+    where Model: FluentKit.Model, Field: FieldRepresentable
+{
+    return .init(lhs, .lessThanOrEqual, rhs)
+}
+
+// not sure about these operators yet
+/*
```
Comment (tanner0101) [inline Sources/FluentKit/Query/QueryBuilder.swift]: Yeah these don't make sense for field to field comparisons. We can delete.

File: Sources/FluentKit/Query/QueryBuilder.swift [hunk 3]
```diff
@@ -267,7 +276,18 @@ public final class QueryBuilder<Model>
...
+    {
+        return self.filter(
+            .field(path: filter.lhsPath, schema: Alias.alias, alias: nil),
+            filter.method,
+            .field(path: filter.rhsPath, schema: Alias.alias, alias: nil)
+        )
+    }
+
+    @discardableResult
+    public func filter<Joined>(_ alias: Joined.Type, _ filter: ModelValueFilter<Joined>) -> Self
```
Comment (JaapWijnen) [inline Sources/FluentKit/Query/QueryBuilder.swift]: Not sure if I'm doing the right thing here, haven't really wrapped my head around `Joined` and `Alias`

File: Sources/FluentKit/Query/QueryBuilder.swift [hunk 4]
```diff
@@ -276,11 +296,27 @@ public final class QueryBuilder<Model>
...
+    @discardableResult
+    public func filter<Joined>(_ alias: Joined.Type, _ filter: ModelFieldFilter<Joined, Joined>) -> Self
+        where Joined: FluentKit.Model
+    {
+        return self.filter(
+            .field(path: filter.lhsPath, schema: Joined.schema, alias: nil),
+            filter.method,
+            .field(path: filter.rhsPath, schema: Joined.schema, alias: nil)
+        )
+    }
```
Comment (JaapWijnen) [inline Sources/FluentKit/Query/QueryBuilder.swift]: Same as above.

File: Sources/FluentKit/Query/QueryBuilder.swift [hunk 5]
```diff
@@ -276,11 +296,27 @@ public final class QueryBuilder<Model>
...
     
     @discardableResult
     public func filter<Value>(_ field: KeyPath<Model, Field<Value>>, _ method: DatabaseQuery.Filter.Method, _ value: Value) -> Self {
         return self.filter(Model.key(for: field), method, value)
     }
+
+    @discardableResult
+    public func filter<Value>(_ lhsField: KeyPath<Model, Field<Value>>, _ method: DatabaseQuery.Filter.Method, _ rhsField: KeyPath<Model, Field<Value>>) -> Self {
+        return self.filter(Model.key(for: lhsField), method, Model.key(for: rhsField))
+    }
```
Comment (JaapWijnen) [inline Sources/FluentKit/Query/QueryBuilder.swift]: This is now defined with regard to the generic `Model` of `QueryBuilder` is that the correct way to do it?

File: Sources/FluentKit/Query/QueryBuilder.swift [hunk 6]
```diff
@@ -293,9 +329,23 @@ public final class QueryBuilder<Model>
...
 
+    @discardableResult
+    public func filter(_ lhsFieldName: String, _ method: DatabaseQuery.Filter.Method, _ rhsFieldName: String) -> Self {
+        return self.filter(
+            .field(path: [lhsFieldName], schema: Model.schema, alias: nil),
+            method,
+            .field(path: [rhsFieldName], schema: Model.schema, alias: nil)
+        )
+    }
+
```
Comment (JaapWijnen) [inline Sources/FluentKit/Query/QueryBuilder.swift]: Same problem here with the generic `Model` type

File: Sources/FluentSQL/SQLQueryConverter.swift [hunk 7]
```diff
@@ -228,6 +228,16 @@ public struct SQLQueryConverter {
                     right: self.value(value)
                 )
             }
+        case .field(let lhsField, let method, let rhsField):
+            switch method {
+                // possible overrides for other methods
```
Comment (JaapWijnen) [inline Sources/FluentSQL/SQLQueryConverter.swift]: Is anything specific needed here?

File: Sources/FluentKit/Query/QueryBuilder.swift (hunk 3 above)
Comment (tanner0101) [reply Sources/FluentKit/Query/QueryBuilder.swift]: Yeah, this seems right :) Alias and Joined are basically exact the same except that Alias has a different table name. You can see an example of how that works in practice here: https://github.com/vapor/fluent-kit/blob/master/Sources/FluentBenchmark/FluentBenchmarker.swift#L1258

File: Sources/FluentSQL/SQLQueryConverter.swift (hunk 7 above)
Comment (tanner0101) [reply Sources/FluentSQL/SQLQueryConverter.swift]: i think we're fine to just return `SQLBinaryExpression` directly here.

**Comment by tanner0101** [comment]
We should also have a test for a field to field filter with a join. LMK if you want me to help with that. 👍

**Comment by JaapWijnen** [comment]
@tanner0101 Implemented your feedback, just have to think of a representable test now... any inspiration?

**Comment by JaapWijnen** [comment]
added a test

**Pull request merged**